src/validate_async.cc

```diff
@@ -11,115 +11,153 @@ using Nan::AsyncWorker;
+                    v8refract = refract2v8(result, { true, DRAFTER_SERIALIZE_JSON });
+                }
+
+                if (!parse_err_code) {
+                    resolver->Resolve(Nan::GetCurrentContext(), v8refract);
                 } else {
-                    argv[1] = Nan::Null();
+                    resolver->Reject(Nan::GetCurrentContext(), v8refract);
+                }
+                v8::Isolate::GetCurrent()->RunMicrotasks();
+                return;
+            } else if (callback) {
+                Local<Value> argv[2];
+
+                switch (parse_err_code) {
+                    case DRAFTER_EUNKNOWN:
+                        argv[0] = Nan::Error("Parser: Unknown Error");
+                        argv[1] = Nan::Null();
+                        break;
+                    case DRAFTER_EINVALID_INPUT:
+                        argv[0] = Nan::Error("Parser: Invalid Input");
+                        argv[1] = Nan::Null();
+                        break;
+                    case DRAFTER_EINVALID_OUTPUT:
+                        argv[0] = Nan::Error("Parser: Invalid Output");
+                        argv[1] = Nan::Null();
+                        break;
+                    default:
+                        argv[0] = Nan::Null();
+                        if (result) {
+                            argv[1] = refract2v8(result, { true, DRAFTER_SERIALIZE_JSON });
+                        } else {
+                            argv[1] = Nan::Null();
+                        }
                 }
+                callback->Call(2, argv);
+                return;
+            }
+            Nan::ThrowTypeError("not handled OKCallback");
         }
 
-        callback->Call(2, argv);
-    }
-
-    virtual ~ValidateWorker() {
-        if (result) {
-            drafter_free_result(result);
-            result = nullptr;
+        virtual ~ValidateWorker()
+        {
+            if (result) {
+                drafter_free_result(result);
+                result = nullptr;
+            }
+
+            if (sourceData) {
+                delete sourceData;
+                sourceData = nullptr;
+            }
         }
+    };
 
-        if (sourceData) {
-            delete sourceData;
-            sourceData = nullptr;
-        }
-    }
-};
+} // namespace
+
+static bool isLastParamCallback(const Nan::FunctionCallbackInfo<v8::Value>& info)
+{
+    return info[info.Length() - 1]->IsFunction();
 }
 
-NAN_METHOD(protagonist::Validate) {
+NAN_METHOD(protagonist::Validate)
+{
     Nan::HandleScope scope;
 
-    // Check arguments
-    if (info.Length() != 2 && info.Length() != 3) {
-        Nan::ThrowTypeError(
-            "wrong number of arguments, `validate(string, options, callback)` "
-            "expected");
+    Nan::Persistent<v8::Promise::Resolver>* persistent = nullptr;
+    Callback* callback = nullptr;
+    v8::Local<v8::Promise::Resolver> resolver;
+
+    size_t optionIndex = 0;
```

Comment:
Same comment as in the parse file. In fact all the other comments in that file applies to this file as well since the code is the same.